src/core/music_score_loaders.cpp

```diff
@@ -0,0 +1,444 @@
+}
+
+void MusicScoreLoader__3_4_2::serializeIndividualBarLabel(std::vector<uint8_t>& data, const MusicScoreBarLabel& barLabel) const {
+	using CalChart::Parser::AppendAndNullTerminate;
+
+	AppendAndNullTerminate(data, barLabel.label);
+}
+
+void MusicScoreLoader__3_4_2::serializeMusicScoreMoment(std::vector<uint8_t>& data, const MusicScoreMoment& moment, const std::unordered_map<const MusicScoreFragment*, uint32_t>& fragmentToIdMap) const {
+	using CalChart::Parser::Append;
+
+	Append(data, (uint32_t)(fragmentToIdMap.at(moment.fragment.get())));
+	Append(data, (uint32_t)(moment.beatAndBar.bar));
+	Append(data, (uint32_t)(moment.beatAndBar.beat));
+}
+
+std::unordered_map<const MusicScoreFragment*, uint32_t> MusicScoreLoader__3_4_2::buildFragmentToIdMap(const MusicScoreDocComponent* docComponent) const {
+	std::unordered_map<const MusicScoreFragment*, uint32_t> returnVal;
+	for (unsigned index = 0; index < docComponent->getNumScoreFragments(); index++) {
+		returnVal.emplace(docComponent->getScoreFragment(index).get(), index + 1);
+	}
+	returnVal.emplace(nullptr, 0);
+	return returnVal;
+}
+
+template <typename EventType>
+void MusicScoreLoader__3_4_2::serializeCollectionOfMusicEvents(std::vector<uint8_t>& data, const CollectionOfMusicScoreEvents<EventType>* eventCollection, const std::unordered_map<const MusicScoreFragment*, uint32_t>& fragmentToIdMap, bool serializeDefaultEvent,
+	const std::function<void(std::vector<uint8_t>&, const EventType&, const std::unordered_map<const MusicScoreFragment*, uint32_t>&)>& eventSerializationFunction) const {
+	using CalChart::Parser::Append;
+
+	if (serializeDefaultEvent) {
+		eventSerializationFunction(data, eventCollection->getDefaultEvent(), fragmentToIdMap);
+	}
+	Append(data, countNumEventsToSerialize<EventType>(eventCollection, fragmentToIdMap));
+	for (auto validFragmentIterator = fragmentToIdMap.cbegin(); validFragmentIterator != fragmentToIdMap.cend(); validFragmentIterator++) {
+		const MusicScoreFragment* fragment = (*validFragmentIterator).first;
+		for (unsigned index = 0; index < eventCollection->getNumEvents(fragment); index++) {
+			auto eventPair = eventCollection->getEvent(fragment, index);
+			serializeMusicScoreMoment(data, eventPair.first, fragmentToIdMap);
+			eventSerializationFunction(data, eventPair.second, fragmentToIdMap);
+		}
+	}
+}
+
+template <typename EventType>
+uint32_t MusicScoreLoader__3_4_2::countNumEventsToSerialize(const CollectionOfMusicScoreEvents<EventType>* eventCollection, const std::unordered_map<const MusicScoreFragment*, uint32_t>& fragmentToIdMap) const {
+	uint32_t totalEvents = 0;
+	for (auto iterator = fragmentToIdMap.cbegin(); iterator != fragmentToIdMap.cend(); iterator++) {
+		totalEvents += eventCollection->getNumEvents((*iterator).first);
+	}
+	return totalEvents;
+}
+
+MusicScoreDocComponent* MusicScoreLoader__3_4_2::loadFromData(const uint8_t*& data, const uint8_t* upperBound) const {
```

Comment:
Why return the raw pointer?  How about instead return std::unique_ptr<MusicScoreDocComponent> ?